### PR TITLE
Applied schizoanalysis: Sanity checking for landscape

### PR DESCRIPTION
**Pull request by liam-fitzgerald**
%sane: sanity checker for the landscape suite of applications

Userspace currently uses certain identifiers as foreign keys, and
expects those foreign keys to exist in a number of locations. 
These foreign key relationships are prone to breaking during OTAs 
and there are enough of them that they rarely get tested for 
manually. %sane is a gall app that will check the validity of
these relationships.

Pokes:
  %off - turn off checking on a reload
  %lax - check on reload, printing errors
  %strict - check on reload, aborting commit if invalid
  %check-print  - run a check now, printing errors
  %check-crash - run a check now, crashing nondeterministically if invalid 
    
Currently validates:
  - Entries in {contact,metadata,group} stores are in sync with 
     their hooks
  - Each group has its associated metadata and contacts
 
 Future validation:
- Subscriptions are correctly setup for hooks
- Graph store integration

### CODE REVIEW

**Comment by matildepark** [comment]
-_-

File: pkg/arvo/app/sane.hoon [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+++  print
+  |=  [pri=@ud =issues]
+  ^+  same
+  %.  same
+  %-  %*(. slog pri pri)
+  issues
+::
+++  xor
+  |*  [a=(set) b=(set)]
+  %-
```
Comment (belisarius222) [inline pkg/arvo/app/sane.hoon:151]: I think this line is short enough for flat-form
```
(~(uni in (~(dif in a) b)) (~(dif in b) a))
```

File: pkg/arvo/app/sane.hoon [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+::      - Entries in {contact,metadata,group} stores are in sync with 
+::         their hooks
+::      - Each group has its associated metadata and contacts
+::
+::    Future validation: 
+::      - Subscriptions are correctly setup for hooks
+::      - Graph store integration
+::    
+::
+
```
Comment (tacryt-socryp) [inline pkg/arvo/app/sane.hoon:30]: empty line

File: pkg/arvo/app/sane.hoon [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+::
+/-  *metadata-store, contacts=contact-store, *group
+/+  default-agent, verb, dbug, resource
+~%  %sane-app  ..is  ~
+|%
++$  card  card:agent:gall
+::
```
Comment (tacryt-socryp) [inline pkg/arvo/app/sane.hoon:37]: No need for `::` in between these lines of `+$` unless one of the declarations spans multiple lines.

File: pkg/arvo/app/sane.hoon [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+  ^-  (quip card _this)
+  :_  this
+  ~[subscribe-to-agent-builds]
+++  on-save  !>(state)
+::
+::
+++  on-load
+  |=  =vase
+  =/  old
+    !<(state-zero vase)
```
Comment (tacryt-socryp) [inline pkg/arvo/app/sane.hoon:73]: Why not put this on a single line? Cleaner.

File: pkg/arvo/app/sane.hoon [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+::
+::
+++  on-poke  
+  |=  [=mark =vase]
+  ^-  (quip card _this)
+  ?.  =(%noun mark)
+    (on-poke:def mark vase)
+  ~!  action
+  =/  act=action
+    !<(action vase)
```
Comment (tacryt-socryp) [inline pkg/arvo/app/sane.hoon:84]: same

File: pkg/arvo/app/sane.hoon [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+  %-  (check-sane:sc act)
+  `this
+::
+::
+++  on-agent  on-agent:def
+::
+++  on-watch  on-watch:def
+::
+++  on-leave  on-leave:def
+
```
Comment (tacryt-socryp) [inline pkg/arvo/app/sane.hoon:96]: blankline

File: pkg/arvo/app/sane.hoon [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+  =/  act=action
+    !<(action vase)
+  ?:  ?=(^mode act)
+    `this(mode act)
+  %-  (check-sane:sc act)
+  `this
+::
+::
+++  on-agent  on-agent:def
+::
```
Comment (tacryt-socryp) [inline pkg/arvo/app/sane.hoon:92]: no need for this `::` separation for single line declarations

File: pkg/arvo/app/sane.hoon [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+++  on-watch  on-watch:def
+::
```
Comment (tacryt-socryp) [inline pkg/arvo/app/sane.hoon:94]: same

File: pkg/arvo/app/sane.hoon [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+    (on-poke:def mark vase)
+  ~!  action
+  =/  act=action
+    !<(action vase)
+  ?:  ?=(^mode act)
+    `this(mode act)
+  %-  (check-sane:sc act)
+  `this
+::
+::
```
Comment (tacryt-socryp) [inline pkg/arvo/app/sane.hoon:90]: no need for two of these lines, one is good enough

File: pkg/arvo/app/sane.hoon [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+  ?.  ?=([%x %bad-path ~] path)
+    (on-peek:def path)
+  ~
+::
+++  on-arvo   
+  |=  [=wire =sign-arvo]
+  ?>  ?=([%rebuilds ~] wire)
+  ?>  ?=([%c %wris *] sign-arvo)
+  =/  ucheck
+    check-type:sc
```
Comment (tacryt-socryp) [inline pkg/arvo/app/sane.hoon:109]: single line

File: pkg/arvo/app/sane.hoon [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+::    metadata, group and graph
+++  store-hook-desync
+  ^-  issues
+  =|  =issues
+  |^
+  =.  issues
+    metadata-desync
+  =.  issues
+    contact-desync
+  issues
```
Comment (tacryt-socryp) [inline pkg/arvo/app/sane.hoon:185]: `::` after this return (of issues)

File: pkg/arvo/app/sane.hoon [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+  ++  metadata-desync
+    ^+  issues
+    =/  groups=(set path)
+      =-  ~(key by -)
+      (scry (jug path md-resource) /y/metadata-store/group-indices)
+    =/  group-syncs
+      (scry (set path) /x/metadata-hook/synced/noun)
+    =/  desyncs=(set path)
+      (xor groups group-syncs)
+    (report-desync %metadata-store desyncs)
```
Comment (tacryt-socryp) [inline pkg/arvo/app/sane.hoon:205]: `::` after this line

File: pkg/arvo/app/sane.hoon [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+  ?~  issues
+    ((print 1 ~[leaf+"Sane!"]) same)
+  %-  (print 3 issues)
+  ?:  =(%check-print check)
+    same
+  =/  failure
+    ~|  %crashing-to-abort-merge
+    (bex (bex 256))
+  same
+::  +store-hook-desync: check desync between store and hookk
```
Comment (tacryt-socryp) [inline pkg/arvo/app/sane.hoon:173]: extra k on the word hook

File: pkg/arvo/app/sane.hoon [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+    =/  desyncs
+      (xor contact-syncs contacts)
+    (report-desync %contact-store desyncs)
+  --
+::  +metadata-group-desync: check desync between metadata and groups
+::
+++  metadata-group-desync
+  ^-  issues
+  =/  groups=(set path)
+    group-store-paths
```
Comment (tacryt-socryp) [inline pkg/arvo/app/sane.hoon:221]: single line

File: pkg/arvo/app/sane.hoon [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+  =/  metadata
```
Comment (tacryt-socryp) [inline pkg/arvo/app/sane.hoon:222]: single line

File: pkg/arvo/app/sane.hoon [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+    metadata-store-paths
+  =/  desyncs
```
Comment (tacryt-socryp) [inline pkg/arvo/app/sane.hoon:224]: single line

File: pkg/arvo/app/sane.hoon [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+      (scry (jug path md-resource) /y/metadata-store/group-indices)
+    =/  group-syncs
+      (scry (set path) /x/metadata-hook/synced/noun)
+    =/  desyncs=(set path)
+      (xor groups group-syncs)
+    (report-desync %metadata-store desyncs)
+  ++  contact-desync
+    ^+  issues
+    =/  contacts
+      contact-store-paths
```
Comment (tacryt-socryp) [inline pkg/arvo/app/sane.hoon:209]: single line

**Review by tacryt-socryp — APPROVED** [review]
Looks excellent, just style nits.

**Comment by matildepark** [comment]
Conflict. Shucks

**Comment by liam-fitzgerald** [comment]
refactored to automatically fix everything by calling `-sane`, should be ready for merge. @tacryt-socryp can I get another pass?

**Review by tacryt-socryp — COMMENTED** [review]
The code looks like it does what it says on the box. But do you think that we should try to "re-add contacts/metadata data" when it's missing, or just remove the group? I'd advocate for removing the group and starting over when that type of desync has occurred.

**Comment by liam-fitzgerald** [comment]
What kind of cases does that cover that simply readding to the hook does not?

**Comment by tacryt-socryp** [comment]
When someone is kicked from a group, they do not lose contacts or metadata, but they do lose the group. This is a single example, but often when partial data is hanging around, it's best to remove it rather than attempt to restore it (because the host's permissions would not allow the client state to be fixed)

**Comment by matildepark** [comment]
OK, merged the candidate back into userspace. Can you resolve these conflicts so we can merge? This has been at large for a few months now.

**Comment by matildepark** [comment]
OK @liam-fitzgerald @tacryt-socryp this conversation seems unsettled. Are we good to merge this? Liam never quite responded.

**Comment by tacryt-socryp** [comment]
@liam-fitzgerald ping, let's discuss this today

**Comment by matildepark** [comment]
So how'd that conversation go?

**Comment by matildepark** [comment]
@tacryt-socryp One last look at this before merge?

**Review by tacryt-socryp — APPROVED** [review]
LGTM

**Pull request merged**